pkg/virt-handler/cgroup/cgroup.go

```diff
@@ -138,6 +163,23 @@ func getCpuSetPath(manager Manager, cpusetFile string) (string, error) {
 	return cpusetStr, nil
 }
 
+// set cpus "cpusList" on the allowed CPUs. Optionally on a subcgroup of
+// the pods control group (if subcgroup != nil).
+func SetCpuSet(manager Manager, subCgroup string, cpusList []int) error {
```

Comment:
The suggested modification does not work since 

func newManagerFromSlicePath(procCgroupBasePath string) (manager Manager, err error) {
        const isRootless = false
        var version CgroupVersion

        // This line isn't needed anymore
        // procCgroupBasePath := filepath.Join(procMountPoint, strconv.Itoa(pid), cgroupStr)
        controllerPaths, err := runc_cgroups.ParseCgroupFile(procCgroupBasePath)

procCgroupBasePath expects a /proc/$pid/cgroup file (and not a cgroup directory).

What can be done is to fake a controllerPaths with only the "cpuset" cgroup, for example. Is that desired?

Reply:
The suggested modification does not work since 

func newManagerFromSlicePath(procCgroupBasePath string) (manager Manager, err error) {
        const isRootless = false
        var version CgroupVersion

        // This line isn't needed anymore
        // procCgroupBasePath := filepath.Join(procMountPoint, strconv.Itoa(pid), cgroupStr)
        controllerPaths, err := runc_cgroups.ParseCgroupFile(procCgroupBasePath)

procCgroupBasePath expects a /proc/$pid/cgroup file (and not a cgroup directory).

What can be done is to fake a controllerPaths with only the "cpuset" cgroup, for example. Is that desired?

Reply:
Attempted to implement this suggestion. Unfortunately it would be necessary to reimplement the code related to construction of "controllerPaths":

 func (v *v2Manager) SetCpuSet(subcgroup string, cpulist []int) error {
-       return setCpuSetHelper(v, subcgroup, cpulist)
+       cgrouppaths := make(map[string]string)
+       subsystemPath := v.Path("cpuset")
+       cpusetpath := strings.ReplaceAll(subsystemPath, "/sys/fs/cgroup/", "/")
+       cgrouppaths[""] = cpusetpath
+       fmt.Printf("SetCpuSet: Path is %s\n", subsystemPath)
+       fmt.Printf("BuiltPaths: %v\n", cgrouppaths)
+       return setCpuSetHelper(v, subcgroup, cpulist, cgrouppaths)
 }

This construction (cgrouppaths == controllerPaths) is incomplete resulting in:

2022-10-14T13:50:07.420149683Z newManagerFromSlicePath cPaths: map[:/ cpuset:/kubepods.slice/kubepods-pod4add640a_0346_4d7c_8bde_094626a92d5a.slice/crio-238a0b570cabbe616371e7f5b90836f5398a25fc204e8787157df72b1e9396f7.scope]
2022-10-14T13:50:07.420172634Z {"component":"virt-handler","kind":"","level":"info","msg":"housekeeping cpu: 2","name":"vmi-fedora-isolated","namespace":"default","pos":"vm.go:2603","timestamp":"2022-10-14T13:50:07.419875Z","uid":"d1e37791-81ea-4dfc-bf7e-bccee5e6c374"}
2022-10-14T13:50:07.420172634Z {"component":"virt-handler","level":"info","msg":"initialized a new cgroup v1 manager successfully. controllerPaths: map[:/ cpuset:/cpuset/kubepods.slice/kubepods-pod4add640a_0346_4d7c_8bde_094626a92d5a.slice/crio-238a0b570cabbe616371e7f5b90836f5398a25fc204e8787157df72b1e9396f7.scope]\n","pos":"cgroup.go:126","timestamp":"2022-10-14T13:50:07.419928Z"}
2022-10-14T13:50:07.420172634Z {"component":"virt-handler","kind":"","level":"error","msg":"Synchronizing the VirtualMachineInstance failed.","name":"vmi-fedora-isolated","namespace":"default","pos":"vm.go:1783","reason":"devices subsystem's path is not defined for this manager","timestamp":"2022-10-14T13:50:07.419958Z","uid":"d1e37791-81ea-4dfc-bf7e-bccee5e6c374"}

So dropping this suggestion for the time being.

Reply:
Firstly, your formatting makes it very hard to read your comments. I recommend using markdown syntax to properly capture code sections etc.

> procCgroupBasePath expects a /proc/$pid/cgroup file (and not a cgroup directory).
> What can be done is to fake a controllerPaths with only the "cpuset" cgroup, for example. Is that desired?

Yes, I think it is.
My problem with the current implementation is that we're re-implementing runc's logic. They already have a declarative interface in which all we need to do is say "I want this and that cpu set for cgroup X" and it will be done - we should be using it.

On the long run, runc may introduce better support / optimizations / etc that we would get for free once updating. Re-inventing the same wheel would always keep us behind and increase the chance of having bugs and problems.

What I suggest:
1. Create a function that can return a sub-cgroup manager.
2. Use that manager to modify stuff in the runc way, with minimum intervention from us (ideally we won't touch any cgroup files directly).

Reply:
@iholder-redhat

Unfortunately sub-cgroups are not supported by runc. Therefore the creation of a sub-cgroup is necessary (therefore it won't be possible to touch any cgroup files directly). BTW, current code does touch cgroup files
directly:

```
func getCpuSetPath(manager Manager, cpusetFile string) (string, error) {
        cpuSubsystemPath, err := manager.GetBasePathToHostSubsystem("cpuset")
        if err != nil {
                return "", err
        }

        cpuset, err := os.ReadFile(filepath.Join(cpuSubsystemPath, cpusetFile))
        if err != nil {
                return "", err
        }

        cpusetStr := strings.TrimSpace(string(cpuset))
        return cpusetStr, nil
}
```

Are you OK with integrating the code as is now and then push everything (including creation of sub-cgroups) to runc afterwards?

Reply:
Firstly, your formatting makes it very hard to read your comments. I recommend using markdown syntax to properly capture code sections etc.

If you can suggest precisely what should be improved, that would be helpful.

Reply:
> the creation of a sub-cgroup is necessary

That's right, this part would have to be done manually by us.
But, after the child cgruop is created, we can make a manager to it (with `newV1Manager` or `newV2Manager` via providing the relevant controller paths). Then we can delegate the task of setting a cpuset to this newly created manager.

WDYT?

Reply:
The functions to create cgroups from pid's make use of /proc/$pid/cgroup (the list of hierarchies attached). Since for a NewManagerFromPath function there won't be any task to refer to, one has to determine which subsystems are enabled for  a given cgroup only from its path. This adds more cgroup specific code to KubeVirt than the few lines you are trying to integrate:

```
// set cpus "cpusList" on the allowed CPUs. Optionally on a subcgroup of
// the pods control group (if subcgroup != nil).
func setCpuSetHelper(manager Manager, subCgroup string, cpusList []int) error {
        subSysPath, err := manager.GetBasePathToHostSubsystem("cpuset")
        if err != nil {
                return err
        }

        if subCgroup != "" {
                subSysPath = filepath.Join(subSysPath, subCgroup)
        }

        wVal := strings.Trim(strings.Replace(fmt.Sprint(cpusList), " ", ",", -1), "[]")

        return runc_cgroups.WriteFile(subSysPath, "cpuset.cpus", wVal)
}
```

Unless you have an additional suggestion?